### PR TITLE
Create a non-IID CIFAR10 datasets for cross-silo experiments

### PR DESCRIPTION
**Pull request by JYWa**
Similar to the federated version of CIFAR100, we follow the methods in [this paper](https://arxiv.org/pdf/1909.06335.pdf). Currently, we assume that there are total 10 clients, each of which has 5000 images.

### CODE REVIEW

**Comment by AdvaitGadhikar** [comment]
@googlebot I signed it

**Comment by AdvaitGadhikar** [comment]
@googlebot I consent.

**Review by nightldj — CHANGES_REQUESTED** [review]
Can we split this into two: one for the dataset, and a second one for the training loop? 

Also, we should not make changes to the optimization/ folder. Let us try to merge the python binary in to fedopt_guide/. 

Thanks!

**Comment by JYWa** [comment]
> Can we split this into two: one for the dataset, and a second one for the training loop?
> 
> Also, we should not make changes to the optimization/ folder. Let us try to merge the python binary in to fedopt_guide/.
> 
> Thanks!

Hi Zheng,

Thanks for the suggestions! Now this PR only contains the changes about the dataset in `utils`.

File: utils/datasets/cifar10_dataset.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    Loads the train dataset into a non iid distribution over clients using the
+    sampling method based on LDA, taken from this paper:
+    Measuring the Effects of Non-Identical Data Distribution for
+    Federated Visual Classification (https://arxiv.org/pdf/1909.06335.pdf).
+    '''
+    (train_images, train_labels), (test_images, test_labels) = datasets.cifar10.load_data()
+    client_train_dataset = collections.OrderedDict()
+    client_test_dataset = collections.OrderedDict()
+
+    # for each class in the dataset, there are 10 for CIFAR-10
```
Comment (nightldj) [inline utils/datasets/cifar10_dataset.py:43]: This comment is unclear. Maybe something like "The following structure will store the image for each of ten classes..."

File: utils/datasets/cifar10_dataset.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    idx_batch = [[] for _ in range(num_clients)]
+
+    for k in range(num_classes):
+        '''
+        This will form the final dataset dictionary for each client
+        For each class(label) sample in a proportion for each client. 
+        The proportion is determined by a dirichlet distribution.
+        '''
```
Comment (nightldj) [inline utils/datasets/cifar10_dataset.py:51]: Use '#' for multi-line comments. https://google.github.io/styleguide/pyguide.html#385-block-and-inline-comments

**Review by nightldj — CHANGES_REQUESTED** [review]

File: utils/datasets/cifar10_dataset.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    num_classes: int = 10, 
+    alpha: float = 1, 
+    train_client_batch_size: int = 20, 
+    test_client_batch_size: int = 100):
+    '''
+    Loads the train dataset into a non iid distribution over clients using the
+    sampling method based on LDA, taken from this paper:
+    Measuring the Effects of Non-Identical Data Distribution for
+    Federated Visual Classification (https://arxiv.org/pdf/1909.06335.pdf).
+    '''
```
[the remark this answers is not in the document]
Comment (JYWa) [reply utils/datasets/cifar10_dataset.py]: Revised. Thanks!

File: utils/datasets/cifar10_dataset.py (hunk 2 above)
Comment (JYWa) [reply utils/datasets/cifar10_dataset.py]: Revised. Thanks!

File: utils/datasets/cifar10_dataset.py (hunk 1 above)
Comment (JYWa) [reply utils/datasets/cifar10_dataset.py]: Revised. Thanks!

**Review by nightldj — APPROVED** [review]

File: utils/datasets/cifar10_dataset.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+import numpy as np
+import tensorflow as tf
+import tensorflow_federated as tff
+from tensorflow.keras import datasets, layers, models
+
+CIFAR_SHAPE = (32, 32, 3)
+TOTAL_FEATURE_SIZE = 32 * 32 * 3
+NUM_EXAMPLES_PER_CLIENT = 5000
+
+def load_cifar10_federated(
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:27]: It would be great to give an optional random seed governing any randomness contained, in order to make things fully reproducible.

I think this can be done via numpy.random.set_state. I'm okay with it in a follow-up, but I think it should be done before experiments start in earnest.

File: utils/datasets/cifar10_dataset.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+  Loads the train dataset into a non iid distribution over clients using the
+  sampling method based on LDA, taken from this paper:
+  Measuring the Effects of Non-Identical Data Distribution for
+  Federated Visual Classification (https://arxiv.org/pdf/1909.06335.pdf).
+
+  Args:
+    num_clients: An integer specifing the total number of clients.
+    num_classes: An integer representing the number of lables in the dataset.
+    alpha: A float controling the data heterogeneity among clients.
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:43]: nit: This is the parameter of a symmetric Dirichlet distribution on NUM_CLASSES.

File: utils/datasets/cifar10_dataset.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+import tensorflow_federated as tff
+from tensorflow.keras import datasets, layers, models
+
+CIFAR_SHAPE = (32, 32, 3)
+TOTAL_FEATURE_SIZE = 32 * 32 * 3
+NUM_EXAMPLES_PER_CLIENT = 5000
+
+def load_cifar10_federated(
+    num_clients: int = 10, 
+    num_classes: int = 10,
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:29]: This should be a fixed constant, since it's CIFAR-10.

File: utils/datasets/cifar10_dataset.py [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+    test_client_batch_size: A float representing the batch size during
+      test.
+
+  Returns:
+    A tuple of `tff.simulation.ClientData` representing unpreprocessed
+    train data and test data.
+  '''
+  (train_images, train_labels), (test_images, test_labels) \
+      = datasets.cifar10.load_data()
+  client_train = collections.OrderedDict()
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:55]: maybe something a bit more descriptive, like "train_clients" (in  order to indicate that this contains the data of multiple clients)

File: utils/datasets/cifar10_dataset.py [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+  client_test = collections.OrderedDict()
+
+  # The following structure will store the image index for each client
+  idx_batch = [[] for _ in range(num_clients)]
+
+  for k in range(num_classes):
+    # For each class(label) sample in a proportion for each client. 
+    # The proportion is determined by a dirichlet distribution.
+    label_k = np.where(train_labels.squeeze()==k)[0]
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:64]: I think this is a no-op, since train_labels is just an ndarray of shape (5000,)

File: utils/datasets/cifar10_dataset.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+    np.random.shuffle(label_k)
+    proportion = np.random.dirichlet(alpha*np.ones((num_clients,)))
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:66]: This is a different approach from Hsu et al., 2019. This partitions each label space according to some categorical distribution over clients, instead of making each client have a categorical distribution over labels.

I think the end result is similar, but I don't think it's identical. One important distinction: This approach can create massively unbalanced clients. Is there a specific reason to do the sampling in this manner?

Comment (zcharles8) [reply utils/datasets/cifar10_dataset.py:66]: For posterity: Appendix B of https://arxiv.org/pdf/2003.08082.pdf describes how to do the LDA client partitioning in a balanced way, when sampling examples without replacement. I think we should be doing the partitioning in that way.

File: utils/datasets/cifar10_dataset.py [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+    proportion = proportion*(label_k.shape[0])
+    proportion = np.cumsum(proportion).astype(int)
+    split_labels = np.split(label_k, proportion)
+      
+    idx_batch = [idx_j + splitk.tolist() for idx_j, splitk \
+       in zip(idx_batch, split_labels)]
+
+  num_per_client_test = int((len(test_labels) \
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:74]: len(test_labels) is a fixed constant

File: utils/datasets/cifar10_dataset.py [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+      // (num_clients*test_client_batch_size))*test_client_batch_size)
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:75]: I don't quite understand why you're using division without remainder here. I don't think it's important to make sure that clients have number of examples divisible by the batch size.

File: utils/datasets/cifar10_dataset.py [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+    num_classes: int = 10, 
+    alpha: float = 1, 
+    train_client_batch_size: int = 20, 
+    test_client_batch_size: int = 100):
+  '''Construct a synthetic federated CIFAR-10 from its centralized version
+
+  Loads the train dataset into a non iid distribution over clients using the
+  sampling method based on LDA, taken from this paper:
+  Measuring the Effects of Non-Identical Data Distribution for
+  Federated Visual Classification (https://arxiv.org/pdf/1909.06335.pdf).
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:38]: One important note: That paper creates balanced clients. I think that as a first pass, we should just restrict to making balanced clients as well. If we think it's necessary, perhaps we can add functionality for unbalanced clients later (perhaps drawing the number of examples per client via a separate Dirichlet with parameter beta)

File: utils/datasets/cifar10_dataset.py [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+        // train_client_batch_size) * train_client_batch_size
+    x_train = x_train[:train_samples_per_client]
+    y_train = y_train[:train_samples_per_client]
+    
+    data = collections.OrderedDict((('image', x_train), ('label', y_train)))
+    test_idx_start = i*num_per_client_test
+    test_idx_end = (i+1)*num_per_client_test       
+    test_data = collections.OrderedDict(
+        (('image', test_images[test_idx_start:test_idx_end]), 
+        ('label', test_labels[test_idx_start:test_idx_end].astype('int64').squeeze())))
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:91]: It seems that we're simply partitioning the test examples in order across clients. Couldn't we use LDA here too?

**Review by zcharles8 — CHANGES_REQUESTED** [review]
Hi all. Thanks for the work on your end. I left a number of comments, but my biggest issue is that this seems to be creating categorical distributions over clients, whereas we should be creating categorical distributions over classes.

**Comment by AdvaitGadhikar** [comment]
Hi Everyone, 

Thank you for your constructive comments about the sampling method! I have tried to incorporate the changes mentioned.
1. I am now sampling a multinomial for the label distribution on each client from the Dirichlet distribution and using this to assign samples to every client.
2. I noticed that if the number of samples at each client is not divisible by the batch size, the training loop gives a shape error for the last batch, hence I have kept the division without remainder
3. I have also taken into account the other suggested edits.
I request you all to review it once and let us know what you think!

File: utils/datasets/cifar10_dataset.py [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+      = datasets.cifar10.load_data()
+  train_clients = collections.OrderedDict()
+  test_clients = collections.OrderedDict()
+
+  proportion_clients_train = []
+  proportion_clients_test = []
+  # each column is a distribution over classes for the client at that index
+  for i in range(num_clients):
+      proportion = np.random.dirichlet(nit*np.ones(NUM_CLASSES,))
+      train_proportion = (5000*proportion).astype(int)
```
Comment (nightldj) [inline utils/datasets/cifar10_dataset.py:61]: Let us replace "5000" with NUM_EXAMPLES_PER_CLIENT?

**Review by nightldj — COMMENTED** [review]
LGTM to me. Thanks!

File: utils/datasets/cifar10_dataset.py [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+    np.random.shuffle(test_label_k)
+    test_split_labels = np.split(test_label_k, cumsum_client_labels_test[:, k])
+    test_idx_batch = [idx_j + splitk.tolist() for idx_j, splitk \
+       in zip(test_idx_batch, test_split_labels)]
+
+  for i in range(num_clients):
+    client_name = str(i)
+    x_train = train_images[np.array(train_idx_batch[i])]
+    y_train = train_labels[np.array(train_idx_batch[i])].astype('int64').squeeze()
+    train_samples_per_client = (x_train.shape[0] \
```
Comment (nightldj) [inline utils/datasets/cifar10_dataset.py:91]: The dependency on `client_batch_size` is strange. Add some comments if it is absolutely necessary.

File: utils/datasets/cifar10_dataset.py [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+  proportion_clients_train = np.array(proportion_clients_train)
+  proportion_clients_test = np.array(proportion_clients_test)
+
+  cumsum_client_labels_train = proportion_clients_train.cumsum(axis=0)
+  cumsum_client_labels_test = proportion_clients_test.cumsum(axis=0)
+
+  for k in range(NUM_CLASSES):
+    train_label_k = np.where(train_labels==k)[0]
+    np.random.shuffle(train_label_k)
+    train_split_labels = np.split(train_label_k, cumsum_client_labels_train[:, k])
```
[the remark this answers is not in the document]
Comment (zcharles8) [reply utils/datasets/cifar10_dataset.py:77]: I ran an example. It doesn't incur an error, but the sampling proportion was such that the latter clients did not receive any examples. This procedure should be altered to factor in the fact that there are exactly 5000 examples available. I think this should be done in a sampling-based way. Once a label is exhausted, a client can simply update its sampling probabilities to remove the exhausted class.

**Review by zcharles8 — CHANGES_REQUESTED** [review]
Currently, there's no guarantee that we will sample all the examples of a given label, or more problematic, that a client will get to sample any. Rather than doing the sampling by taking the sampling probabilities an d multiplying by NUM_EXAMPLES_PER_CLIENT, I think they should actually be sampled sequentially. This allows us to update the sampling once a label is exhausted.

Please let me know if you have any questions about the above.

File: utils/datasets/cifar10_dataset.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (AdvaitGadhikar) [reply utils/datasets/cifar10_dataset.py:77]: Thank you for your comments! 
Yes, I understand what you mean, there could potentially be two cases, one where cum_sum_client_labels_train[:, k] sums up to more than 5000 and another where it sums up to less than 5000. In the latter case, this method would not account for the remaining samples in that class. Is this the problem you are referring to, just want to be sure?
Sampling them sequentially will exhaust all the samples in every class, I will work on that change

**Comment by AdvaitGadhikar** [comment]
Hi All,

I have made the suggested edits and now the sampling process at each client is sequential and it ensures that each client has 5000 samples with a label distribution according to a multinomial sampled from a Dirichlet distribution. I have also removed the dependence on the batch size for now.

File: utils/datasets/cifar10_dataset.py [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+import tensorflow as tf
+import tensorflow_federated as tff
+from tensorflow.keras import datasets, layers, models
+
+CIFAR_SHAPE = (32, 32, 3)
+TOTAL_FEATURE_SIZE = 32 * 32 * 3
+NUM_EXAMPLES_PER_CLIENT = 5000
+TEST_SAMPLES_PER_CLIENT = 1000
+NUM_CLASSES = 10
+NUM_CLIENTS = 10
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:28]: Side note: I don't think it's necessary to restrict this, but this is fine for now. We can revisit if necessary.

File: utils/datasets/cifar10_dataset.py [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+  train_client_samples = [[] for _ in range(NUM_CLIENTS)]
+  test_client_samples = [[] for _ in range(NUM_CLIENTS)]
+  '''
+  sample 5000 images for the current client according 
+  to the multinomial without replacement
+  '''
+  train_count = np.zeros(NUM_CLASSES).astype(int)
+  test_count = np.zeros(NUM_CLASSES).astype(int)
+
+  for k in range(NUM_CLASSES):
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:82]: should this be NUM_CLIENTS?

File: utils/datasets/cifar10_dataset.py [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+        
+    for i in range(NUM_EXAMPLES_PER_CLIENT):
+      sampled_label = np.argwhere(np.random.multinomial(1, train_multinomial_vals[k,:])==1)[0][0]
+      train_client_samples[k].append(train_indices[sampled_label, train_count[sampled_label]])
+      train_count[sampled_label] += 1
+      if train_count[sampled_label] == NUM_EXAMPLES_PER_CLIENT:
+        train_multinomial_vals[:, sampled_label] = 0
+        train_multinomial_vals = train_multinomial_vals / train_multinomial_vals.sum(axis=1)[:, None]
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:90]: Nice!

File: utils/datasets/cifar10_dataset_test.py [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+import tensorflow as tf
+
+from utils.datasets import cifar10_dataset
+
+
+def _compute_length_of_dataset(ds):
+  return ds.reduce(0, lambda x, _: x + 1)
+
+
+class DatasetTest(tf.test.TestCase):
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset_test.py:26]: It might be worth having a test to ensure that the number of clients is 10, and that (at least one of) the clients have the correct number of batches.

**Review by zcharles8 — APPROVED** [review]

File: utils/datasets/cifar10_dataset.py (hunk 18 above)
Comment (AdvaitGadhikar) [reply utils/datasets/cifar10_dataset.py:82]: I think it should be NUM_CLASSES and we would have to change NUM_EXAMPLES_PER_CLIENT if we want to change the number of clients.

**Review by nightldj — APPROVED** [review]

**Comment by nightldj** [comment]
Merged in https://github.com/google-research/federated/commit/cb2518e0da738b8bf8d9c145459aa717a71133aa. Thank you for the contribution!
